[RFC] oldtest: set helpfile for runtime/doc/tags in build dir

blueyed opened the pull request:
This is better practice in general, and allows to remove the "helptags
ALL" hacks.

Ref: https://github.com/neovim/neovim/issues/8824
Ref: https://github.com/neovim/neovim/commit/f1b67c3453c

blueyed committed:
oldtest: use runtime from build

This is better practice in general, and allows to remove the "helptags
ALL" hacks.

Ref: https://github.com/neovim/neovim/issues/8824
Ref: https://github.com/neovim/neovim/commit/f1b67c3453c

blueyed commented:
Ok, that would need to copy/configure the runtime, which is a bit too much I guess.

The problem here is that Vim ships `runtime/doc/tags`, while Neovim does not - it seems to be forgotten in a98a6996c maybe (at least it is not explicitly mentioned).

blueyed committed:
src/nvim/testdir/runtest.vim: set helpfile

justinmk commented:
> The problem here is that Vim ships `runtime/doc/tags`, while Neovim does not - it seems to be forgotten in [a98a699](https://github.com/neovim/neovim/commit/a98a6996c291b3e300d27b791eded9eed333d677) maybe (at least it is not explicitly mentioned).

wasn't forgotten, we intentionally decided not to include generated files in the tree. Doing that isn't much better of a "hack" than simply running `:helptags ALL` ...

blueyed commented:
Ok.  I can see the reason for it.

I think using `set helpfile=…` is a better/smaller diff then to get Vim's behavior in tests.

justinmk commented:
> I think using `set helpfile=…` is a better/smaller diff then to get Vim's behavior in tests.

That changes the assumed state though. Should be done in specific tests maybe.

blueyed committed:
add build dir to rtp

blueyed commented:
> That changes the assumed state though.

The state is different however.

What about adding the runtime from the build dir?  (latest commit)

blueyed committed:
bufenter_spec: do not create runtime/doc/tags as a side-effect

blueyed committed:
fix test/functional/autocmd/filetype_spec.lua

blueyed committed:
fix test/functional/provider/python3_spec.lua

blueyed committed:
TEST: test/functional/provider/python3_spec.lua:109: comment :help help

blueyed committed:
fix test/functional/ex_cmds/help_spec.lua

blueyed committed:
TEMP: no AppVeyor

blueyed committed:
TEMP: Travis: single job

blueyed committed:
fix "ruby provider" test

blueyed committed:
debug hanging ruby-provider test

blueyed committed:
ruby-provider: use command

blueyed committed:
Makefile: fix dependencies with regard to helptags

- use the file as the main target to avoid unnecessary triggering
- use "make oldtest" on Travis to ensure it gets built
- functionaltests also need to depend on it

blueyed commented:
Should also use `make oltest` on AppVeyor then - but having it built via functionaltest before might be good.  Code ref: https://github.com/neovim/neovim/blob/46cb85691c18f99af70de65d0168e972b1dc57dd/ci/build.ps1#L137-L143

blueyed committed:
cmake: add ${GENERATED_HELP_TAGS} to FUNCTIONALTEST_PREREQS

blueyed committed:
squash! Makefile: fix dependencies with regard to helptags

Revert change to functionaltest*: handled via cmake

blueyed committed:
Revert "TEST: test/functional/provider/python3_spec.lua:109: comment :help help"

This reverts commit 14c22c1cab32c2581a1281e5c98cfebd2d68a8a5.

blueyed committed:
Revert "ruby-provider: use command"

This reverts commit 198e399c352b61c154db5c5cc402c18e8ac03049.

blueyed committed:
Revert "debug hanging ruby-provider test"

This reverts commit cb036afa8b869eca61ba18b80cf10c07b8075d1c.

blueyed committed:
Revert "TEMP: no AppVeyor"

This reverts commit 565c18feb27d845a7553ecd2f5ed8ea568f330f9.

blueyed committed:
Revert "TEMP: Travis: single job"

This reverts commit 28065c8b94b33e2cd033271e7ff5e4c40cc5e4a3.

blueyed committed:
fixup! fix test/functional/provider/python3_spec.lua

blueyed committed:
set rtp in functional.helpers.nvim_set

blueyed committed:
Revert "set rtp in functional.helpers.nvim_set"

This reverts commit 384e519c78249fe59c1e3945828bd8d276d2ba3b.

blueyed commented:
There are failures with adding build/runtime to rtp always: https://travis-ci.org/neovim/neovim/jobs/574180741#L6426

What do you think?
Reverted 384e519c7 for now.

blueyed commented on test/functional/autocmd/bufenter_spec.lua:
```diff
@@ -33,9 +33,10 @@ describe('autocmd BufEnter', function()
     command("split normal")
     command("wincmd j")
-    command("helptags runtime/doc")
     command("help")
```
Maybe we could have a `TestHelp` command instead that would adjust `rtp` then before calling `help`?
But it's also OK maybe having it done 5 times manually?

justinmk replied:
a function would be useful, if there's some reason we can't just do it once in `clear()`.

blueyed replied:
https://github.com/neovim/neovim/commit/384e519c78249fe59c1e3945828bd8d276d2ba3b did not work - i.e. that would need other test adjustments then.

justinmk replied:
No, I mean just do whatever you're already doing, except in `clear()`.

blueyed replied:
So the nearest `clear()` to the tests that need it, or in the default/global clear?  (the latter would be the same I guess with regard to unrelated/different failures then)

justinmk replied:
I was thinking global clear, but sounds like that is too much trouble. It's better to explicit I guess, anyway (i.e. `helpers.add_builddir_to_rtp()`).

blueyed replied:
Good idea, done.

justinmk commented on src/nvim/testdir/test_taglist.vim:
```diff
@@ -89,8 +89,11 @@ func Test_tagfiles()
   let tf = tagfiles()
+  " Nvim: expectation(s) based on tags in build dir (added to &rtp).
+  "       Filter out the (non-existing) '../../../runtime/doc/tags'.
+  call filter(tf, 'filereadable(v:val)')
```
is this still needed? why not do `helptags ALL` once?

justinmk commented on CMakeLists.txt:
```diff
@@ -541,6 +541,7 @@ add_subdirectory(test/includes)
 add_subdirectory(config)
 add_subdirectory(test/functional/fixtures)  # compile test programs
 add_subdirectory(runtime)
+get_directory_property(GENERATED_HELP_TAGS DIRECTORY runtime DEFINITION GENERATED_HELP_TAGS)
```
is this redefining `GENERATED_HELP_TAGS`? should it use a different var name, e.g. `GENERATED_HELP_TAGS_DIR`

blueyed replied:
It is pulling it in from the subdirectory.
Could have a different name, but it is the same after all.

blueyed committed:
helpers.add_builddir_to_rtp

blueyed committed:
rebuild
